### PR TITLE
feat(GAQ): Add Redis Sentinel Support for Global Async Queries

### PR DESCRIPTION
**Pull request by nsivarajan**
### SUMMARY

This PR introduces a feature that allows the Async Query Manager to use a configured cache backend through the new **GLOBAL_ASYNC_QUERIES_CACHE_BACKEND** setting. To maintain backward compatibility, the existing **GLOBAL_ASYNC_QUERIES_REDIS_CONFIG** setting is still supported but will be deprecated in the future. Additionally, this update _introduces support for Redis Sentinel caching_, which eliminates the single point of failure associated with the previous standalone Redis configuration.

References taken from this Idea https://github.com/apache/superset/discussions/28839

### TESTING INSTRUCTIONS

Configure **GLOBAL_ASYNC_QUERIES_CACHE_BACKEND** in _config.py_ or _superset_config.py_ with the appropriate properties:
Set CACHE_TYPE to "RedisCache" for Redis cache backend.
Set CACHE_TYPE to "RedisSentinelCache" for Redis Sentinel cache backend.
Set CACHE_TYPE to None to fall back on the previous GLOBAL_ASYNC_QUERIES_REDIS_CONFIG.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by nsivarajan** [comment]
HI @villebro , Could you please review this PR? It was created based on the idea discussed https://github.com/apache/superset/discussions/28839

However, I recently noticed that there is a new SIP in the design phase to modernize the framework around the same topic: [#29839](https://github.com/apache/superset/issues/29839) . I'm not sure if this PR will still add value before the new proposal is rolled out, but I would appreciate your feedback.

**Comment by villebro** [comment]
@nsivarajan thanks for the PR - I think this PR is welcomed even if there's ongoing work to modernize the async task framework. Looks generally good, but let me do a proper review shortly (either today or tomorrow)

**Comment by nsivarajan** [comment]
@villebro Thanks for the feedback! Looking forward to your review.

**Comment by nsivarajan** [comment]
Hi @villebro,

I wanted to follow up on this PR. I understand you're busy, but I was hoping to get your review when you have a moment. Please let me know if there's any additional information I can provide or if there are any specific areas you'd like me to address.

Thanks again for your time and feedback!

**Comment by rusackas** [comment]
@nsivarajan thank you for your patience in the meantime... good to see that CI is passing and there's no conflicts still, at least! :D

File: tests/unit_tests/async_events/async_query_manager_tests.py [hunk 1]
```diff
@@ -26,17 +27,27 @@
...
 JWT_TOKEN_SECRET = "some_secret"
 JWT_TOKEN_COOKIE_NAME = "superset_async_jwt"
 
+# Define the cache backends once as mocks
+cache_backends = {
+    "RedisCacheBackend": mock.Mock(spec=RedisCacheBackend),
+    "RedisSentinelCacheBackend": mock.Mock(spec=RedisSentinelCacheBackend),
+    "redis.Redis": mock.Mock(spec=redis.Redis),
+}
+
```
Comment (villebro) [inline tests/unit_tests/async_events/async_query_manager_tests.py:44]: Could this loop be replaced by `pytest.mark.parametrize`? It feels more idiomatic (unless I'm missing some reason why it needs to be implemented like this..)

**Review by villebro — COMMENTED** [review]
Thanks so much @nsivarajan for this beautiful PR (the test coverage here is great)! Sorry for the long review time - there was a lot to take in, and I may need to do a new round after I get some feedback on these comments. Please let me know what you think, and I'll do my best to do a second (and hopefully final) pass shortly.

File: superset/async_events/async_query_manager.py [hunk 2]
```diff
@@ -230,14 +257,35 @@ def submit_chart_data_job(
...
+                    {
+                        key.decode("utf-8"): value.decode("utf-8")
+                        for key, value in event_data.items()
+                    },
+                )
+                for event_id, event_data in results
+            ]
+            return (
+                [] if not decoded_results else list(map(parse_event, decoded_results))
+            )
```
[the remark this answers is not in the document]
Comment (nsivarajan) [reply superset/async_events/async_query_manager.py:280]: RedisCache and RedisSentinelCache typically don't support built-in decoding like decode_responses. They return data in bytes, which needs to be manually decoded into the desired formats (e.g., strings).

https://github.com/pallets-eco/flask-caching/blob/master/src/flask_caching/backends/rediscache.py#L155

File: superset/async_events/async_query_manager.py [hunk 3]
```diff
@@ -16,18 +16,27 @@
 # under the License.
 import logging
 import uuid
-from typing import Any, Literal, Optional
+from typing import Any, Dict, Literal, Optional, Union
```
[the remark this answers is not in the document]
Comment (nsivarajan) [reply superset/async_events/async_query_manager.py:19]: Thank you for pointing that out.  I'll make sure to update the code accordingly.

File: tests/unit_tests/async_events/async_query_manager_tests.py (hunk 1 above)
Comment (nsivarajan) [reply tests/unit_tests/async_events/async_query_manager_tests.py:44]: Yes, that's a great suggestion. I'll update the code with `pytest.mark.parametrize` and commit the changes.

Comment (nsivarajan) [reply tests/unit_tests/async_events/async_query_manager_tests.py]: Using `pytest.mark.parametrize` with `mock` from `unittest` caused compatibility issues. Switching to `parameterized.expand` resolved these issues and successfully covered the test cases, leading to a successful CI run. I’ll revisit this case later for further investigation.

Comment (villebro) [reply tests/unit_tests/async_events/async_query_manager_tests.py]: @nsivarajan I believe it's generally suggested to use `pytest_mock` with `pytest` test cases. I don't want to drag out this review process more than necessary, but I think the cleanest solution would be one with `pytest.mark.parametrize` and `pytest_mock`. LMKWYT, but in the meantime let me re-review the functional parts.

File: tests/integration_tests/tasks/async_queries_tests.py [hunk 4]
```diff
@@ -37,18 +43,37 @@
 )
 from tests.integration_tests.test_app import app
 
+# Define the cache backends once
+cache_backends = {
+    "RedisCacheBackend": mock.Mock(spec=RedisCacheBackend),
+    "RedisSentinelCacheBackend": mock.Mock(spec=RedisSentinelCacheBackend),
+    "redis.Redis": mock.Mock(spec=redis.Redis),
+}
+
```
Comment (villebro) [inline tests/integration_tests/tasks/async_queries_tests.py:52]: I believe these are now redundant as we have the parametrized tests?
```suggestion
# Define the cache backends once
cache_backends = {
    "RedisCacheBackend": mock.Mock(spec=RedisCacheBackend),
    "RedisSentinelCacheBackend": mock.Mock(spec=RedisSentinelCacheBackend),
    "redis.Redis": mock.Mock(spec=redis.Redis),
}

```

File: tests/unit_tests/async_events/async_query_manager_tests.py [hunk 5]
```diff
@@ -17,26 +17,37 @@
...
 JWT_TOKEN_SECRET = "some_secret"
 JWT_TOKEN_COOKIE_NAME = "superset_async_jwt"
 
+# Define the cache backends once as mocks
+# cache_backends = {
+#    "RedisCacheBackend": mock.Mock(spec=RedisCacheBackend),
+#    "RedisSentinelCacheBackend": mock.Mock(spec=RedisSentinelCacheBackend),
+#    "redis.Redis": mock.Mock(spec=redis.Redis),
+# }
+
```
Comment (villebro) [inline tests/unit_tests/async_events/async_query_manager_tests.py:44]: Same here:
```suggestion
```

**Review by villebro — COMMENTED** [review]
A few minor observations during second review round

File: superset/async_events/async_query_manager.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (villebro) [reply superset/async_events/async_query_manager.py:282]: Got it, I hadn't noticed we were using `redis.Redis`, I thought we were using `RedisCache` from `flask_caching` here. Thanks for clarifying.

File: tests/unit_tests/async_events/async_query_manager_tests.py (hunk 5 above)
Comment (nsivarajan) [reply tests/unit_tests/async_events/async_query_manager_tests.py]: Thanks, I'll get this removed.

File: tests/integration_tests/tasks/async_queries_tests.py (hunk 4 above)
Comment (nsivarajan) [reply tests/integration_tests/tasks/async_queries_tests.py]: Thanks, I'll get this removed.

File: tests/unit_tests/async_events/async_query_manager_tests.py (hunk 5 above)
Comment (nsivarajan) [reply tests/unit_tests/async_events/async_query_manager_tests.py]: Updated in last commit

File: tests/integration_tests/tasks/async_queries_tests.py (hunk 4 above)
Comment (nsivarajan) [reply tests/integration_tests/tasks/async_queries_tests.py]: Handled in last commit

File: tests/unit_tests/async_events/async_query_manager_tests.py (hunk 1 above)
Comment (nsivarajan) [reply tests/unit_tests/async_events/async_query_manager_tests.py]: Thanks, @villebro , for the review. I agree that using the standard approach is ideal. However, I've encountered continuous failures when combining `pytest.mark.parametrize` with `mock`, which suggests there might be compatibility issues. After some research, I found references that indicate potential conflicts when using pytest features with unittest.TestCase subclasses and mock (see [pytest documentation](https://docs.pytest.org/en/stable/how-to/unittest.html#pytest-features-in-unittest-testcase-subclasses), [SeleniumBase issue](https://github.com/seleniumbase/SeleniumBase/issues/395), and [pytest issue](https://github.com/pytest-dev/pytest/issues/541#issuecomment-609956951)).

I noticed that our repository already uses `parameterized.expand` in similar cases, so I opted for it here to ensure the tests run smoothly.

Would it be possible to proceed with this PR using `parameterized.expand`? I'll continue investigating the issues with `pytest.mark.parametrize` and `mock` and plan to submit a follow-up PR if I find a viable solution.

Comment (villebro) [reply tests/unit_tests/async_events/async_query_manager_tests.py]: Thanks for putting in the extra effort on these tests @nsivarajan ! Let's not let this derail your PR - I can also take a stab at migrating the remaining components to `pytest` (seems like a fun coffee break challenge).

File: superset/async_events/async_query_manager.py [hunk 6]
```diff
@@ -55,13 +66,32 @@ def parse_event(event_data: tuple[str, dict[str, Any]]) -> dict[str, Any]:
...
+        prefix, last = entry_id[:-1], int(entry_id[-1])
         return prefix + str(last + 1)
     except Exception:  # pylint: disable=broad-except
-        return redis_id
+        return entry_id
+
+
+def get_cache_backend(
+    config: dict[str, Any],
+) -> Union[RedisCacheBackend, RedisSentinelCacheBackend, redis.Redis]:  # type: ignore
```
Comment (villebro) [inline superset/async_events/async_query_manager.py:80]: nit: the new type annotations support this simpler format:
```suggestion
) -> RedisCacheBackend | RedisSentinelCacheBackend | redis.Redis:  # type: ignore
```

File: superset/async_events/async_query_manager.py [hunk 7]
```diff
@@ -73,7 +103,7 @@ class AsyncQueryManager:
 
     def __init__(self) -> None:
         super().__init__()
-        self._redis: redis.Redis  # type: ignore
+        self._cache: Optional[BaseCache] = None
```
Comment (villebro) [inline superset/async_events/async_query_manager.py:106]: Same here:
```suggestion
        self._cache: BaseCache | None = None
```

**Review by villebro — APPROVED** [review]
LGTM, thanks for all the hard work here! I've left a few extra type annotation tips, but they're not blocking (we've already done enough iterating here). I'll add a new task to our 5.0 project board to remove the old config. If you feel up for it, please open a PR with the change that removes the old config option so we can merge it when the 5.0 breaking window is opened up.

**Comment by nsivarajan** [comment]
Thanks, @villebro, for the feedback and type annotation tips!

I appreciate your guidance throughout this process. I'll go ahead and open a new PR to remove the old config option and will reference the task once it's created on the project board, so it’ll be ready when the 5.0 breaking window opens up. Looking forward to wrapping this up and moving on to the next steps!

**Comment by villebro** [comment]
Thanks @nsivarajan, I really appreciate the help! Here's the project card: https://github.com/orgs/apache/projects/345?pane=issue&itemId=78081192

**Pull request merged**